### ISSUE TITLE
maven-metadata.xml file outdated

### ISSUE THREAD

**Issue by elruwen** [dzikoysk/reposilite#1821]
### What happened?

Hi!

I am using reposilite as a mirror of my jfrog artifactory maven repo for performance reasons.

So I configured it as a mirror repo and I am caching the artifacts.

Every now and then the maven-metadata.xml file does not get updated. Let say of my artifact mygroup:foo I got version 1.2.3-4 in the repo and then I push the artifactory 1.2.3-5.

I can see in the UI the 1.2.3-5 artifact, but the maven-metadata.xml is not listing that version. The maven-metadata.xml in artifactory has the correct version. I know that artifactory is sometimes a bit slow to propagate changes, but the problem even happens after hours.
When I delete from reposilite the maven-metadata.xml and run on my project `mvn -U versions:use-latest-versions` reposilite grabs the correct file.

How does the maven-metadata.xml file gets updated?
Is reposilite checking it for updates each time run  `mvn -U versions:use-latest-versions` ?

Version I use: 3.1.0
My config:
```
{
  "frontend": {
    "id": "reposilite-repository",
    "title": "XXX",
    "description": "XXX",
    "organizationWebsite": "XXX",
    "organizationLogo": "XXX",
    "icpLicense": ""
  },
  "authentication": {
    "ldap": {
      "enabled": false,
      "hostname": "ldap.domain.com",
      "port": 389,
      "baseDn": "dc=company,dc=com",
      "searchUserDn": "cn=reposilite,ou=admins,dc=domain,dc=com",
      "searchUserPassword": "reposilite-admin-secret",
      "userAttribute": "cn",
      "userFilter": "(&(objectClass=person)(ou=Maven Users))",
      "userType": "PERSISTENT"
    }
  },
  "statistics": {
    "resolvedRequestsInterval": "MONTHLY"
  },
  "web": {
    "forwardedIp": "X-Forwarded-For"
  },
  "maven": {
    "repositories": [
      {
        "id": "releases",
        "visibility": "PUBLIC",
        "redeployment": false,
        "preserveSnapshots": false,
        "storageProvider": {
          "type": "fs",
          "quota": "90%",
          "mount": ""
        },
        "proxied": [
          {
            "reference": "https://repo1.maven.org/maven2/",
            "store": true,
            "allowedGroups": [],
            "connectTimeout": 3,
            "readTimeout": 15,
            "httpProxy": ""
          },
          {
            "reference": "https://jfrog url",
            "store": true,
            "allowedGroups": [],
            "connectTimeout": 3,
            "readTimeout": 15,
            "authorization": {
              "password": "XXX",
              "login": "XXX"
            },
            "httpProxy": ""
          }
        ]
      }
    ]
  }
}
```

### Reposilite version

3.x

**Comment by dzikoysk** [comment]
The intention behind caching is to provide a possibility to go fully offline/independent from mirrored repo at some point, so we're storing all files - even the metadata files. That's the main purpose of this feature, but it also has some drawbacks like the one you've just noticed. 

I see the issue tho. It could be covered by some new option added to the caching policy that'd state that we need to prioritize metadata files over local content.

**Comment by elruwen** [comment]
> The intention behind caching is to provide a possibility to go fully offline/independent from mirrored repo at some point, so we're storing all files - even the metadata files. That's the main purpose of this feature, but it also has some drawbacks like the one you've just noticed.
> 
> I see the issue tho. It could be covered by some new option added to the caching policy that'd state that we need to prioritize metadata files over local content.

I had a bit of a deeper look at the calls maven does. I was hoping it sets the Cache-Control header a bit smart, but it always sets it to no-cache and in newer maven versions even to no-cache, no-store, no matter what file is being requested.  I was hoping to get no-cache, no-store only for the maven-metadata.xml.

Maybe something like this:
{
"pattern" : "**/maven-metadata.xml",
"cacheMode" : "check_upstream_ignore_failure"
}

So basically if a file matches, a call to upstream with the http header If-Modified-Since is being made. If upstream is not available, gracefully read the file from the local cache.

Other cache modes:
- default, current behaviour
- check_upstream - if upstream is not reachable, fail the request - no need to implement it at the moment imho

**Comment by dzikoysk** [comment]
It'll be available in the next release:

![obraz](https://github.com/dzikoysk/reposilite/assets/4235722/8e0496ed-c9bc-4cb9-afd8-84f6a5b58e7d)

**Comment by elruwen** [comment]
Thank you soooo much!

**Comment by elruwen** [comment]
I had a look at the changes.
I was also playing around locally a bit, but I only was able to start. I was doing a GET request to upstream with a If-Modified-Since header. I simply use the last-modifed timestamp from the local file. So I would still do a request, but I would only grab the file if it has changed.

I am curious if you thought about that approach?

**Comment by dzikoysk** [comment]
I think it'd complicate the impl, and it doesn't really provide any extra value for users. In the end, you still need to make a request to an external source, so it'll take as much time as a request that just downloads the metadata file again.
